pkg/rc/replication_controller.go

```diff
@@ -0,0 +1,243 @@
+			return util.Errorf("Not enough nodes to meet desire: %d replicas desired, %d current, %d eligible", rc.ReplicasDesired, len(current), len(eligible))
+		}
+
+		for i := 0; i < toSchedule; i++ {
+			err := rc.schedule(possible[i])
+			if err != nil {
+				return err
+			}
+		}
+	} else if len(current) > rc.ReplicasDesired {
+		toUnschedule := len(current) - rc.ReplicasDesired
+		rc.logger.NoFields().Infof("Need to unschedule %d nodes out of %s", toUnschedule, current)
+
+		for i := 0; i < toUnschedule; i++ {
+			err := rc.unschedule(current[i])
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		rc.logger.NoFields().Debugln("Taking no action")
+	}
+
+	return nil
+}
+
+func (rc *replicationController) eligibleNodes() ([]string, error) {
+	return rc.scheduler.EligibleNodes(rc.Manifest, rc.NodeSelector)
+}
+
+func (rc *replicationController) CurrentNodes() ([]string, error) {
+	selector := rc.PodLabels.AsSelector()
+
+	pods, err := rc.podApplicator.GetMatches(selector, labels.POD)
+	if err != nil {
+		return []string{}, nil
+	}
+
+	// ID will be something like <nodename>/<podid>.
+	// We just want the nodename.
+	result := make([]string, len(pods))
+	for i, node := range pods {
+		parts := strings.SplitN(node.ID, "/", 2)
+		// If it so happens that node.ID contains no "/",
+		// then parts[0] is the entire string, so we're OK
+		result[i] = parts[0]
+	}
+	return result, nil
+}
+
+func (rc *replicationController) forEachLabel(node string, f func(id, k, v string) error) error {
+	id := node + "/" + rc.Manifest.ID()
+
+	// user-requested labels.
+	for k, v := range rc.PodLabels {
+		if err := f(id, k, v); err != nil {
+			return err
+		}
+	}
+	// our reserved labels.
+	if err := f(id, podIdLabel, rc.Manifest.ID()); err != nil {
+		return err
+	}
+	return f(id, rcIdLabel, rc.Id.String())
+}
+
+func (rc *replicationController) schedule(node string) error {
+	// First, schedule the new pod.
+	intentPath := kp.IntentPath(node, rc.Manifest.ID())
+	rc.logger.NoFields().Infof("Scheduling on %s", intentPath)
+	_, err := rc.kpStore.SetPod(intentPath, rc.Manifest)
+	if err != nil {
+		return err
+	}
+
+	// Then we set the labels.
+	// TODO: It could be the case that these SetLabel fail.
+	// In this case, we've scheduled a pod but currentNodes() will never be aware of it.
+	// We could consider cleaning the now-orphaned pod up?
+	return rc.forEachLabel(node, func(id, k, v string) error {
+		return rc.podApplicator.SetLabel(labels.POD, id, k, v)
+	})
+}
+
+func (rc *replicationController) unschedule(node string) error {
+	intentPath := kp.IntentPath(node, rc.Manifest.ID())
+	rc.logger.NoFields().Infof("Uncheduling from %s", intentPath)
+
+	err := rc.forEachLabel(node, func(id, k, _ string) error {
+		return rc.podApplicator.RemoveLabel(labels.POD, id, k)
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = rc.kpStore.DeletePod(intentPath)
+	return err
+}
+
+func difference(subtractFrom, subtractThese []string) []string {
```

Comment:
maybe should use `StringSet.Difference()` from pkg/labels/set.go? Will mean you need to use set types instead of string slices which maybe we don't want if the node ordering matters?